module/zstd/zstd.c

```diff
@@ -0,0 +1,720 @@
+	case ZSTD_KMEM_POOL:
+		zstd_mempool_free(z);
+		break;
+	case ZSTD_KMEM_DCTX:
+		mutex_exit(&zstd_dctx_fallback.barrier);
+		break;
+	default:
+		break;
+	}
+}
+
+/* Allocate fallback memory to ensure safe decompression */
+static void __init
+create_fallback_mem(struct zstd_fallback_mem *mem, size_t size)
+{
+	mem->mem_size = size;
+	mem->mem = vmem_zalloc(mem->mem_size, KM_SLEEP);
+	mutex_init(&mem->barrier, NULL, MUTEX_DEFAULT, NULL);
+}
+
+/* Initialize memory pool barrier mutexes */
+static void __init
+zstd_mempool_init(void)
+{
+	zstd_mempool_cctx = (struct zstd_pool *)
+	    kmem_zalloc(ZSTD_POOL_MAX * sizeof (struct zstd_pool), KM_SLEEP);
+	zstd_mempool_dctx = (struct zstd_pool *)
+	    kmem_zalloc(ZSTD_POOL_MAX * sizeof (struct zstd_pool), KM_SLEEP);
+
+	for (int i = 0; i < ZSTD_POOL_MAX; i++) {
+		mutex_init(&zstd_mempool_cctx[i].barrier, NULL,
+		    MUTEX_DEFAULT, NULL);
+		mutex_init(&zstd_mempool_dctx[i].barrier, NULL,
+		    MUTEX_DEFAULT, NULL);
+	}
+}
+
+/* Initialize zstd-related memory handling */
+static int __init
+zstd_meminit(void)
+{
+	zstd_mempool_init();
+
+	/*
+	 * Estimate the size of the fallback decompression context.
+	 * The expected size on x64 with current ZSTD should be about 160 KB.
+	 */
+	create_fallback_mem(&zstd_dctx_fallback,
+	    P2ROUNDUP(ZSTD_estimateDCtxSize() + sizeof (struct zstd_kmem),
+	    PAGESIZE));
+
+	return (0);
+}
+
+/* Release object from pool and free memory */
+static void __exit
+release_pool(struct zstd_pool *pool)
+{
+	mutex_destroy(&pool->barrier);
+	vmem_free(pool->mem, pool->size);
+	pool->mem = NULL;
+	pool->size = 0;
+}
+
+/* Release memory pool objects */
+static void __exit
+zstd_mempool_deinit(void)
+{
+	for (int i = 0; i < ZSTD_POOL_MAX; i++) {
+		release_pool(&zstd_mempool_cctx[i]);
+		release_pool(&zstd_mempool_dctx[i]);
+	}
+
+	kmem_free(zstd_mempool_dctx, ZSTD_POOL_MAX * sizeof (struct zstd_pool));
+	kmem_free(zstd_mempool_cctx, ZSTD_POOL_MAX * sizeof (struct zstd_pool));
+	zstd_mempool_dctx = NULL;
+	zstd_mempool_cctx = NULL;
+}
+
+extern int __init
+zstd_init(void)
+{
+	/* Set pool size by using maximum sane thread count * 4 */
+	pool_count = (boot_ncpus * 4);
+	zstd_meminit();
+
+	/* Initialize kstat */
+	zstd_ksp = kstat_create("zfs", 0, "zstd", "misc",
+	    KSTAT_TYPE_NAMED, sizeof (zstd_stats) / sizeof (kstat_named_t),
+	    KSTAT_FLAG_VIRTUAL);
+	if (zstd_ksp != NULL) {
+		zstd_ksp->ks_data = &zstd_stats;
+		kstat_install(zstd_ksp);
+	}
+
+	return (0);
+}
+
+extern void __exit
+zstd_fini(void)
+{
+	/* Deinitialize kstat */
+	if (zstd_ksp != NULL) {
+		kstat_delete(zstd_ksp);
+		zstd_ksp = NULL;
+	}
+
+	/* Release fallback memory */
+	vmem_free(zstd_dctx_fallback.mem, zstd_dctx_fallback.mem_size);
+	mutex_destroy(&zstd_dctx_fallback.barrier);
+
+	/* Deinit memory pool */
+	zstd_mempool_deinit();
+}
+
+#if defined(_KERNEL)
+module_init(zstd_init);
+module_exit(zstd_fini);
+
+ZFS_MODULE_DESCRIPTION("ZSTD Compression for ZFS");
+ZFS_MODULE_LICENSE("Dual BSD/GPL");
```

Comment:
isnt that in conflict wit hthe CDDL license header?

Reply:
It's just wrong.
I think this is caused by our old folder structure not being reflected correctly in this code...
Any input @c0d3z3r0 

should be something like:
ZFS_MODULE_DESCRIPTION("ZSTD Compression for ZFS");
ZFS_MODULE_LICENSE("CDDL");
ZFS_LIBRARY_LICENSE("Dual BSD/GPL");

Reply:
We had some discussion on that on one of the earlier PRs. Since zstd itself is Dual BSD/GPL, we cannot make `zstd.c` CCDL, can we? I think it is the same "problem" like spl/Linux being GPL which CCDL is not compatible to. Have a look at `module/os/linux/spl/spl-generic.c` or `module/lua/lapi.c` for more examples.

Maybe @behlendorf and @Cyan4973 can give us a hint on doing licensing correct here?

Reply:
No that's bullshit.
Library use is fine.

Reply:
You can't add a CDDL header to BSD/GPL, so please make up your mind what it is.

Reply:
what is "bullshit"?

Reply:
> You can't add a CDDL header to BSD/GPL, so please make up your mind what it is.

well, that's what I said... no idead why you're attacking me now...

"Since zstd itself is Dual BSD/GPL, we cannot make zstd.c CCDL, can we?"

Reply:
> No that's bullshit.
> Library use is fine.

mmh thats not true. this is true it would be dynamically linked. but its not. its compiled into our code and becomes part of the whole libary. so we do not use just a api interface. we compile the whole code into it. that makes a huge difference

Reply:
The entire point of it being BSD licensed, is that there are no restrictions on what you link it to.

Reply:
But, it seems to make most sense to keep it all CDDL so it isn't a strange separate part of ZFS. Although, previously it was part of the core zfs kernel module, not its own separate module, so maybe there are advantages to doing BSD licensed.

Reply:
yes. i think we need to change the license header to BSD style to be compliant with zstd code since its deep linked with it

Reply:
I think changing the `zstd.c` header to the BSD style license would be best.  Using the BSD license for everything in this kmod provides us the most flexibility which we may want in the future.  I'd suggest we update the headers in `zstd/zstd.c` and `zstd/include/*.h` to all use the 3-clause BSD license (the headers currently use the 2-clause variant).

Reply:
The dual license allow users to opt for _any_ one of them.
So yes, it's possible to label this implementation as "BSD license" only.

Reply:
@behlendorf @Cyan4973 thanks for your feedback, I'll change that and push in the next minutes

Reply:
and making it BSD licensed means it's compatible with CDDL. unlike GPL.

Reply:
> well, that's what I said... no idead why you're attacking me now...
> 
> "Since zstd itself is Dual BSD/GPL, we cannot make zstd.c CCDL, can we?"

I said it at the time hard basically 3 different licences, which is wrong anyway. But I rest my case...